lighthouse-core/gather/driver/wait-for-condition.js

```diff
@@ -0,0 +1,420 @@
+ */
+
+/**
+ * Returns a promise that resolves immediately.
+ * Used for placeholder conditions that we don't want to start waiting for just yet, but still want
+ * to satisfy the same interface.
+ * @return {{promise: Promise<void>, cancel: function(): void}}
+ */
+function waitForNothing() {
+  return {promise: Promise.resolve(), cancel() {}};
+}
+
+/**
+ * Returns a promise that resolve when a frame has been navigated.
+ * Used for detecting that our about:blank reset has been completed.
+ * @param {LH.Gatherer.FRProtocolSession} session
+ * @return {CancellableWait}
+ */
+function waitForFrameNavigated(session) {
+  /** @type {(() => void)} */
+  let cancel = () => {
+    throw new Error('waitForFrameNavigated.cancel() called before it was defined');
+  };
+
+  const promise = new Promise((resolve, reject) => {
+    session.once('Page.frameNavigated', resolve);
+    cancel = () => {
+      session.off('Page.frameNavigated', resolve);
+      reject(new Error('Wait for navigated cancelled'));
+    };
+  });
+
+  return {promise, cancel};
+}
+
+/**
+ * Returns a promise that resolve when a frame has a FCP.
+ * @param {LH.Gatherer.FRProtocolSession} session
+ * @param {number} pauseAfterFcpMs
+ * @param {number} maxWaitForFcpMs
+ * @return {CancellableWait}
+ */
+function waitForFcp(session, pauseAfterFcpMs, maxWaitForFcpMs) {
+  /** @type {(() => void)} */
+  let cancel = () => {
+    throw new Error('waitForFcp.cancel() called before it was defined');
+  };
+
+  /** @type {Promise<void>} */
+  const promise = new Promise((resolve, reject) => {
+    const maxWaitTimeout = setTimeout(() => {
+      reject(new LHError(LHError.errors.NO_FCP));
+    }, maxWaitForFcpMs);
+    /** @type {NodeJS.Timeout|undefined} */
+    let loadTimeout;
+
+    /** @param {LH.Crdp.Page.LifecycleEventEvent} e */
+    const lifecycleListener = e => {
+      if (e.name === 'firstContentfulPaint') {
+        loadTimeout = setTimeout(() => {
+          resolve();
+          cancel();
+        }, pauseAfterFcpMs);
+      }
+    };
+
+    session.on('Page.lifecycleEvent', lifecycleListener);
+
+    let canceled = false;
+    cancel = () => {
+      if (canceled) return;
+      canceled = true;
+      session.off('Page.lifecycleEvent', lifecycleListener);
+      maxWaitTimeout && clearTimeout(maxWaitTimeout);
+      loadTimeout && clearTimeout(loadTimeout);
+      reject(new Error('Wait for FCP canceled'));
+    };
+  });
+
+  return {
+    promise,
+    cancel,
+  };
+}
+
+/**
+ * Returns a promise that resolves when the network has been idle (after DCL) for
+ * `networkQuietThresholdMs` ms and a method to cancel internal network listeners/timeout.
+ * @param {LH.Gatherer.FRProtocolSession} session
+ * @param {NetworkRecorder} networkMonitor
+ * @param {number} networkQuietThresholdMs
+ * @return {CancellableWait}
+ */
+function waitForNetworkIdle(session, networkMonitor, networkQuietThresholdMs) {
```

Comment:
extracting `session` and `networkMonitor` to parameters are the only substantive changes in this PR